bluemira/equilibria/coils/_coil.py

```diff
@@ -0,0 +1,521 @@
+    """
+    CoilType Enum
+    """
+
+    PF = auto()
+    CS = auto()
+    NONE = auto()
+
+
+class CoilNumber:
+    """
+    Coil naming-numbering utility class. Coil naming convention is not enforced here.
+    """
+
+    __PF_counter: int = 1
+    __CS_counter: int = 1
+    __no_counter: int = 1
+
+    @staticmethod
+    def generate(ctype: CoilType) -> int:
+        """
+        Generate a coil name based on its type and indexing if specified. If no index is
+        specified, an encapsulated global counter assigns an index.
+
+        Parameters
+        ----------
+        coil: Any
+            Object to name
+
+        Returns
+        -------
+        name: str
+            Coil name
+        """
+        if ctype == CoilType.NONE:
+            idx = CoilNumber.__no_counter
+            CoilNumber.__no_counter += 1
+        elif ctype == CoilType.CS:
+            idx = CoilNumber.__CS_counter
+            CoilNumber.__CS_counter += 1
+        elif ctype == CoilType.PF:
+            idx = CoilNumber.__PF_counter
+            CoilNumber.__PF_counter += 1
+
+        return idx
+
+
+class Coil(CoilFieldsMixin):
+    """
+    Coil Object
+
+    For used with PF/CS/passive coils
+
+    Parameters
+    ----------
+    x: float
+        Coil geometric centre x coordinate [m]
+    z: float
+        Coil geometric centre z coordinate [m]
+    dx: float
+        Coil radial half-width [m] from coil centre to edge (either side)
+    dz: float
+        Coil vertical half-width [m] from coil centre to edge (either side)
+    name: str
+        The name of the coil
+    ctype: Union[str, CoilType]
+        Type of coil as defined in CoilType
+    current: float (default = 0)
+        Coil current [A]
+    j_max: float
+        Maximum current density in the coil [A/m^2]
+    b_max: float
+        Maximum magnetic field at the coil [T]
+    discretisation: float
+        discretise the coil (between 0 and 1)
+    n_turns: int
+        Number of turns
+
+    """
+
+    def __init__(
+        self,
+        x: float,
+        z: float,
+        dx: float = None,
+        dz: float = None,
+        name: Optional[str] = None,
+        ctype: Union[str, CoilType] = CoilType.NONE,
+        current: float = 0,
+        j_max: float = np.nan,
+        b_max: float = np.nan,
+        discretisation: float = 1,
+        n_turns: int = 1,
+    ):
+        self._dx = None
+        self._dz = None
+        self._discretisation = 1
+        self._flag_sizefix = None not in (dx, dz)
+
+        self.x = x
+        self.z = z
+        self.dx = dx
+        self.dz = dz
+        self.discretisation = discretisation
+        self.current = current
+        self.j_max = j_max
+        self.b_max = b_max
+        self.ctype = ctype
+        self.name = name
+        self.n_turns = n_turns
+
+        self._number = CoilNumber.generate(self.ctype)
+        if self.name is None:
+            self.name = f"{self._ctype.name}_{self._number}"
+
```

Comment:
The number is a unique identifier its should be used for other things then the name doesnt matter if for whatever reason is duplicated

Reply:
But it is only used for naming?

Reply:
We use the names in eg the region mapping and I would like to not use name as a unique identifier there as it will possibly restrict the user if they have dupe names with the region mapper. In a follow up PR I want to change that to use the unique identifier to link regions with specific coils.